### ISSUE TITLE
Perfomance Testing

### ISSUE THREAD

**Issue by fforootd** [zitadel/zitadel#4424]
## Description

We should define reproducible performance tests for ZITADEL.
This issues aims to outline the goal and task of such an undertaking.

### Login 100 Mio. User a day (aka. 1200 Logins per second)

Tests:

- Login with Authorisation Code Flow (Username+Password)
- Login with Authorisation Code Flow (Passwordless)

### High Number of identities (> 100Mio)

TBB

### High Number of "Active" Instance (> 10000)

TBD

### Actions (TBD)

TBD

## Tools / Infra

- K6 - for the load testing, they even have an [OAuth example](https://k6.io/docs/examples/oauth-authentication/#okta)
- GKE - We should run the ZITADEL Helm Chart with the CockroachDB on top of GKE

Running k6 on K8s looks good https://k6.io/blog/running-distributed-tests-on-k8s/

### Infra Amount

All values are an educated guess based on the data we have at hand.

App Node Group: 50Pods 2CPU, 0.5GB each
Database Node Group: 3 Pods 32CPU, 128GB each
Test Node Group: 1 Pod 16CPU, 64GB Ram

**Acceptance criteria**
- [ ] Define Goal
- [ ] Defined Test
- [ ] Define Infra
- [ ] Create K6s test script for test scenarios
- [ ] Create TF and maybe Terratest code to automated test the whole benchmark @eliobischof 
- [ ] Store Results
- [ ] Improve 😁

**Comment by fforootd** [comment]
@hifabienne I think this is a starting point for the customer inquire we have.

@adlerhurst @livio-a what do you think of these ideas

@eliobischof @thesephirot I think you two would be up for the task

**Comment by adlerhurst** [comment]
From a zitadel perspective it could get interesting to define the test parameters like amount of amount of instances, orgs, users, user types, projects

**Comment by fforootd** [comment]
> From a zitadel perspective it could get interesting to define the test parameters like amount of amount of instances, orgs, users, user types, projects

Makes sense I will quickly adapt some ideas

**Comment by thesephirot** [comment]
I will try an initial setup with k6. for such amounts of users something like https://locust.io/ might do the trick as well

**Comment by fforootd** [comment]
Locust looks fine to me as well 😝

**Comment by adlerhurst** [comment]
Not sure if i would add an additional programming language. Personally i
don’t have much Python knowledge.

**Comment by thesephirot** [comment]
I am with you @adlerhurst , to me it is about the "best" solution for this kind of loadtesting. I am not stuck to a tool, but will evaluate the results.

**Comment by adlerhurst** [comment]
@fforootd Would it make sense to define load profiles as well? Testing the average is fine by me but we all know that there will be a peak when people come back to work

**Comment by fforootd** [comment]
@adlerhurst I agree the 1200/s acts only as starting point to test the system under sustained load.

I think testing for 10000 logins per second is important as well.

**Comment by thesephirot** [comment]
there might be some challenges with k6, currently testing the memory

https://k6.io/docs/examples/data-parameterization/

"...For example, the Cloud service allocates 8GB of memory for every 300 VUs. ..."

**Comment by fforootd** [comment]
> there might be some challenges with k6, currently testing the memory
> 
> https://k6.io/docs/examples/data-parameterization/
> 
> "...For example, the Cloud service allocates 8GB of memory for every 300 VUs. ..."

In my assumption i already planed 64GB so it looks fine from that side, lets try and see how this turns out.

Btw. we do not necessary need a json to store the users. But instead we could generate deterministic usernames and passwords.

**Comment by thesephirot** [comment]
json is very easy to iterate as it has IDS I can pick with the VUs. But we can use any kind of loop

**Comment by eliobischof** [comment]
@hifabienne can we prioritize this?
Especially for the quotas feature, it becomes very relevant #4779 

As a developer, I want to be able to compare my changes impacts on the products performance in easy to execute and reliable iterations, so I can efficiently fix performance problems.

**Comment by eliobischof** [comment]
Closing in favor of #8023

**Comment by elinashoko** [comment]
Thanks for filing this issue. As part of an ongoing backlog cleanup, we're closing issues that have had no activity in over 12 months.

If this is still relevant, let us know in a comment (or reopen the issue) and we'll take another look.

This issue will close in 30 days if there's no further activity.
